### PR TITLE
Tuner cleanup on error

### LINKED ISSUE #13856: Tuner leaves checkpoint files behind when interrupted
## 🐛 Bug

When an error occurs during turning (LR or BS), for example caused by user error in their training_step, the tuner leaves checkpoint files behind in the default_root dir. 

### To Reproduce
```py

import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
        self.learning_rate = 0.1

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        assert False  # simulte an error by user
        return {"loss": loss}

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        auto_lr_find=True
    )
    trainer.tune(model, train_dataloaders=train_data)


if __name__ == "__main__":
    run()

```

### Expected behavior

No files. We should save to a temporary location or use a tempfile.

### Environment

Master 1.7.

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #13856
Tuner algorithms will currently leave behind temp checkpoints if the tuning for some reason fails. PR makes sure these are cleaned up using `try, except, finally` statements even if tuning crashes. 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21162.org.readthedocs.build/en/21162/

### CODE REVIEW

File: src/lightning/pytorch/tuner/batch_size_scaling.py [hunk 1]
```diff
@@ -76,24 +76,27 @@ def _scale_batch_size(
...
+        elif mode == "binsearch":
+            new_size = _run_binary_scaling(trainer, new_size, batch_arg_name, max_trials, params)
 
-    log.info(f"Finished batch size finder, will continue with full run using batch size {new_size}")
+        garbage_collection_cuda()
 
-    __scale_batch_restore_params(trainer, params)
+        log.info(f"Finished batch size finder, will continue with full run using batch size {new_size}")
+    except Exception as e:
+        raise e
```
Comment (Borda) [inline src/lightning/pytorch/tuner/batch_size_scaling.py]:
```suggestion
    except Exception as ex:
        raise ex
```
pls not a single letter :)

File: src/lightning/pytorch/tuner/lr_finder.py [hunk 2]
```diff
@@ -257,40 +257,45 @@ def _lr_find(
...
+        __lr_finder_restore_params(trainer, params)
+
+        if trainer.progress_bar_callback:
+            trainer.progress_bar_callback.enable()
+
+        # Update results across ranks
+        lr_finder.results = trainer.strategy.broadcast(lr_finder.results)
+        lr_finder_finished = True
+    except Exception as e:
+        raise e
```
Comment (Borda) [inline src/lightning/pytorch/tuner/lr_finder.py]:
```suggestion
    except Exception as ex:
        raise ex
```
